test/langtools/tools/javac/processing/model/element/TestUnnamedClass.java

```diff
@@ -0,0 +1,200 @@
+                           RoundEnvironment roundEnv) {
+        if (round == 0) { // Check file from comamnd line
+            checkRoots(roundEnv);
+            generateUnnamed();
+        }
+
+        if (!roundEnv.processingOver()) { // Test generated file(s)
+            checkRoots(roundEnv);
+        }
+
+        round++;
+        return true;
+    }
+
+    private void checkRoots(RoundEnvironment roundEnv) {
+        int checks = 0;
+        for (TypeElement type : typesIn(roundEnv.getRootElements())) {
+            System.out.println("Checking " + type.getQualifiedName());
+            checks++;
+            checkUnnamedClassProperties(type);
+        }
+        if (checks == 0) {
+            messager.printError("No checking done of any candidate unnamed classes.");
+        }
+    }
+
+    private void generateUnnamed() {
+        try {
+            String unnamedSource = """
+            void main() {
+                System.out.println("Nameless, but not voiceless.");
+            }
+            """;
+
+            JavaFileObject outputFile = processingEnv.getFiler().createSourceFile("Nameless");
+            try(Writer w = outputFile.openWriter()) {
+                w.append(unnamedSource);
+            }
+        } catch (java.io.IOException ioe) {
+            throw new RuntimeException(ioe);
+        }
+    }
+
+    /*
+     * From JEP 445 JLS changes:
+     *
+     * "An unnamed class compilation unit implicitly declares a class that satisfies the following
+     * properties:
+     * It is always a top level class.
+     * It is always an unnamed class (it has no canonical or fully qualified name (6.7)).
+     * It is never abstract (8.1.1.1).
+     * It is always final (8.1.1.2).
+     * It is always a member of an unnamed package (7.4.2) and has package access.
+     * Its direct superclass type is always Object (8.1.4).
+     * It never has any direct superinterface types (8.1.5).
+     *
+     * The body of the class contains every ClassMemberDeclaration
+     * from the unnamed class compilation unit. It is not possible for
+     * an unnamed class compilation unit to declare an instance
+     * initializer, static initializer, or constructor.
+     *
+     * It has an implicitly declared default constructor (8.8.9).
+     * All members of this class, including any implicitly declared
+     * members, are subject to the usual rules for member declarations
+     * in a class.
+     *
+     * It is a compile-time error if this class does not declare a candidate main method (12.1.4).
+     */
+    void checkUnnamedClassProperties(TypeElement unnamedClass) {
+        if (unnamedClass.getNestingKind() != NestingKind.TOP_LEVEL) {
+            messager.printError("Unnamed class is not top-level.", unnamedClass);
+        }
+
+        if (!unnamedClass.isUnnamed()) {
+            messager.printError("Unnamed class is _not_ indicated as such.", unnamedClass);
+        }
+
+        if (unnamedClass.getSimpleName().isEmpty()) {
```

Comment:
Presumably the sense of the predicates testing names needs to now be reserved due to subsequent updates to the javax.lang.model support.

Reply:
True, but we should ensure that what we are shipping now does what we claim.